jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/PathResponseListener.java

```diff
@@ -0,0 +1,192 @@
+// ========================================================================
+//
+
+package org.eclipse.jetty.client;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileLock;
+import java.nio.file.Path;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.jetty.client.Response.CompleteListener;
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.util.thread.AutoLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of {@link Response.ContentListener} that produces an {@link Path}
+ * that allows applications to save a file from a response {@link Response}
+ * like curl &lt;URL&gt; -o file.bin does.
+ * <p>
+ * Typical usage is:
+ * <pre>
+ *  httpClient.newRequest(host, port)
+ * .send(new PathResponseListener(Path.of("/tmp/file.bin"));
+ * 
+ *  var request = httpClient.newRequest(host, port);
+ *  CompletableFuture&gt;Path&gt; completable = PathResponseListener.write(request, Path.of("/tmp/file.bin"));
+ * </pre>
+ */
+public class PathResponseListener implements CompleteListener, Response.ContentListener
+{
+    private static final Logger LOG = LoggerFactory.getLogger(InputStreamResponseListener.class);
+    
+    private CompletableFuture<Path> completable = new CompletableFuture<>();
+    private final AutoLock.WithCondition lock = new AutoLock.WithCondition();
+    private final CountDownLatch responseLatch = new CountDownLatch(1);
+    private Path path;
+    private Response response;
+    private Throwable failure;
+    private FileOutputStream fileOut;
+    private FileLock fileLock;
+    
+    public PathResponseListener(Path path) throws FileNotFoundException, IOException
+    {   
+        if (!path.isAbsolute())
+        {
+            throw new FileNotFoundException();
+        }
+        
+        this.path = path;
+        
+        try 
+        {
+            fileOut = new FileOutputStream(this.path.toFile(), true);
+            fileOut.getChannel().truncate(0);
+            fileLock = fileOut.getChannel().lock();
+        }
+        catch (IOException e) 
+        {   
+            if (LOG.isDebugEnabled())
+                LOG.debug("Unable to instantiate object", e);
+            else
+                throw e;
+        }
+    }
+
+    @Override
+    public void onContent(Response response, ByteBuffer content) throws IOException
+    {
+        if (response.getStatus() != HttpStatus.OK_200)
+        {
+            throw new HttpResponseException(String.format("HTTP status code of this response %d", response.getStatus()), response);
+        }
```

Comment:
Move this check to `onHeaders()` so it is done only once. 
See examples in `BufferingResponseListener`.